crates/forge/src/collecting.rs

```diff
@@ -0,0 +1,588 @@
+use crate::{RunnerConfig, RunnerParams, BUILTINS};
+use anyhow::{anyhow, Context, Result};
+use assert_fs::fixture::{FileTouch, PathChild, PathCopy};
+use assert_fs::TempDir;
+use cairo_lang_sierra::program::Program;
+use camino::Utf8PathBuf;
+use rayon::prelude::{IntoParallelRefIterator, ParallelIterator};
+use test_collector::{collect_tests, LinkedLibrary, TestCase};
+use walkdir::WalkDir;
+
+pub struct TestsFromCrate {
+    pub sierra_program: Program,
+    pub test_cases: Vec<TestCase>,
+    pub test_crate_type: TestCrateType,
+}
+
+impl TestsFromCrate {
+    pub fn filter_by_name(&self, filter: &str, exact_match: bool) -> Self {
+        let mut result = vec![];
+        for test in &self.test_cases {
+            if exact_match {
+                if test.name == filter {
+                    result.push(test);
+                }
+            } else if test.name.contains(filter) {
+                result.push(test);
+            }
+        }
+        Self {
+            sierra_program: self.sierra_program.clone(),
+            test_cases: result.into_iter().cloned().collect(),
+            test_crate_type: self.test_crate_type,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum TestCrateType {
+    /// Tests collected from the package
+    Lib,
+    /// Tests collected from the tests folder
+    Tests,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct TestCrate {
+    pub crate_root: Utf8PathBuf,
+    pub crate_name: String,
+    pub crate_type: TestCrateType,
+}
+
+impl TestCrate {
+    pub fn compile_tests(
+        &self,
+        linked_libraries: &[LinkedLibrary],
+        corelib_path: &Utf8PathBuf,
+    ) -> Result<TestsFromCrate> {
+        let (sierra_program, test_cases) = collect_tests(
+            self.crate_root.as_str(),
+            None,
+            &self.crate_name,
+            linked_libraries,
+            Some(BUILTINS.clone()),
+            corelib_path.into(),
+        )?;
+
+        Ok(TestsFromCrate {
+            sierra_program,
+            test_cases,
+            test_crate_type: self.crate_type,
+        })
+    }
+}
+
+pub fn collect_test_crates(
+    package_path: &Utf8PathBuf,
+    package_name: &str,
+    package_source_dir_path: &Utf8PathBuf,
+    temp_dir: &TempDir,
+) -> Result<Vec<TestCrate>> {
+    let tests_dir_path = package_path.join("tests");
+
+    let test_dir_crate = if tests_dir_path.exists() {
+        let lib_path = tests_dir_path.join("lib.cairo");
+        if lib_path.exists() {
+            Some(TestCrate {
+                crate_root: tests_dir_path,
+                crate_name: "tests".to_string(),
+                crate_type: TestCrateType::Tests,
+            })
+        } else {
+            Some(pack_tests_into_one_file(temp_dir, &tests_dir_path)?)
```

Comment:
Why this is done inside of the collecting function?

Reply:
It is kinda a part of test collecting (if we want to return test crates from this function)

Reply:
I disagree, this is detail of compilation not collection, please rename the function or move it to compilation phase.